File: src/test/java/co/nlighten/jsontransform/functions/TransformerFunctionLengthTest.java
package co.nlighten.jsontransform.functions;

import co.nlighten.jsontransform.BaseTest;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;
import java.util.Map;

public class TransformerFunctionLengthTest extends BaseTest {
    @Test
    void autoDetect() {
        var str = "Hello World";
        assertTransformation(str, "$$length:hello world", BigDecimal.valueOf(11));
        assertTransformation(str, "$$length():$", BigDecimal.valueOf(11));
        var arr = new String[]{"Hello", "World"};
        assertTransformation(arr, "$$length:$", BigDecimal.valueOf(2));
        var obj = Map.of("a", "Hello", "b", "World", "c", "foo", "d", "bar");
        assertTransformation(obj, "$$length:$", BigDecimal.valueOf(4));
    }

    @Test
    void stringOnly() {
        var str = "Hello World";
        assertTransformation(str, "$$length(STRING):hello world", BigDecimal.valueOf(11));
        assertTransformation(str, "$$length(STRING):$", BigDecimal.valueOf(11));
        var arr = new String[]{"Hello", "World"};
        assertTransformation(arr, "$$length(STRING):$", null);
        var obj = Map.of("a", "Hello", "b", "World", "c", "foo", "d", "bar");
        assertTransformation(obj, "$$length(STRING):$", null);
    }

    @Test
    void arrayOnly() {
        var str = "Hello World";
        assertTransformation(str, "$$length(ARRAY):hello world", null);
        assertTransformation(str, "$$length(ARRAY):$", null);
        var arr = new String[]{"Hello", "World"};
        assertTransformation(arr, "$$length(ARRAY):$", BigDecimal.valueOf(2));
        var obj = Map.of("a", "Hello", "b", "World", "c", "foo", "d", "bar");
        assertTransformation(obj, "$$length(ARRAY):$", null);
    }

    @Test
    void objectOnly() {
        var str = "Hello World";
        assertTransformation(str, "$$length(OBJECT):hello world", null);
        assertTransformation(str, "$$length(OBJECT):$", null);
        var arr = new String[]{"Hello", "World"};
        assertTransformation(arr, "$$length(OBJECT):$", null);
        var obj = Map.of("a", "Hello", "b", "World", "c", "foo", "d", "bar");
        assertTransformation(obj, "$$length(OBJECT):$", BigDecimal.valueOf(4));
    }

    @Test
    void zeroDefault() {
        assertTransformation(null, "$$length:$", null);
        assertTransformation(null, "$$length(AUTO,true):$", BigDecimal.ZERO);
        assertTransformation(42, "$$length(AUTO,true):$", BigDecimal.ZERO);
        assertTransformation(true, "$$length(AUTO,true):$", BigDecimal.ZERO);
    }
}

File: javascript/json-transform/src/__tests__/functions/TransformerFunctionLength.test.ts
import { describe, test } from "vitest";
import { assertTransformation } from "../BaseTransformationTest";

describe("TransformerFunctionLength", () => {
  test("autoDetect", async () => {
    const str = "Hello World";
    await assertTransformation(str, "$$length:hello world", 11);
    await assertTransformation(str, "$$length():$", 11);
    const arr = ["Hello", "World"];
    await assertTransformation(arr, "$$length:$", 2);
    const obj = { a: "Hello", b: "World", c: "foo", d: "bar" };
    await assertTransformation(obj, "$$length:$", 4);
  });

  test("stringOnly", async () => {
    const str = "Hello World";
    await assertTransformation(str, "$$length(STRING):hello world", 11);
    await assertTransformation(str, "$$length(STRING):$", 11);
    const arr = ["Hello", "World"];
    await assertTransformation(arr, "$$length(STRING):$", null);
    const obj = { a: "Hello", b: "World", c: "foo", d: "bar" };
    await assertTransformation(obj, "$$length(STRING):$", null);
  });

  test("arrayOnly", async () => {
    const str = "Hello World";
    await assertTransformation(str, "$$length(ARRAY):hello world", null);
    await assertTransformation(str, "$$length(ARRAY):$", null);
    const arr = ["Hello", "World"];
    await assertTransformation(arr, "$$length(ARRAY):$", 2);
    const obj = { a: "Hello", b: "World", c: "foo", d: "bar" };
    await assertTransformation(obj, "$$length(ARRAY):$", null);
  });

  test("objectOnly", async () => {
    const str = "Hello World";
    await assertTransformation(str, "$$length(OBJECT):hello world", null);
    await assertTransformation(str, "$$length(OBJECT):$", null);
    const arr = ["Hello", "World"];
    await assertTransformation(arr, "$$length(OBJECT):$", null);
    const obj = { a: "Hello", b: "World", c: "foo", d: "bar" };
    await assertTransformation(obj, "$$length(OBJECT):$", 4);
  });

  test("zeroDefault", async () => {
    await assertTransformation(null, "$$length:$", null);
    await assertTransformation(null, "$$length(AUTO,true):$", 0);
    await assertTransformation(42, "$$length(AUTO,true):$", 0);
    await assertTransformation(true, "$$length(AUTO,true):$", 0);
  });
});
